spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -0,0 +1,371 @@
+    // Reduce scale
+    TypeEvolutionTestCase(DecimalType(Decimal.MAX_INT_DIGITS, 2),
+      DecimalType(Decimal.MAX_INT_DIGITS, 3),
+      Seq("-67.89", "9" * (Decimal.MAX_INT_DIGITS - 2) + ".99")),
+    // Reduce precision
+    TypeEvolutionTestCase(DecimalType(Decimal.MAX_INT_DIGITS, 2),
+      DecimalType(Decimal.MAX_INT_DIGITS - 1, 2),
+      Seq("-67.89", "9" * (Decimal.MAX_LONG_DIGITS - 2) + ".99")),
+    // Reduce precision & scale
+    TypeEvolutionTestCase(DecimalType(Decimal.MAX_INT_DIGITS, 2),
+      DecimalType(Decimal.MAX_INT_DIGITS - 1, 1),
+      Seq("-67.89", "9" * (Decimal.MAX_LONG_DIGITS - 2) + ".99")),
+    // Increase scale more than precision
+    TypeEvolutionTestCase(DecimalType(Decimal.MAX_INT_DIGITS, 2),
+      DecimalType(Decimal.MAX_INT_DIGITS + 1, 4),
+      Seq("-67.89", "9" * (Decimal.MAX_LONG_DIGITS - 2) + ".99"))
+  )
+
+  for {
+    testCase <- unsupportedNonTestCases
+    partitioned <- BOOLEAN_DOMAIN
+  } {
+    test(s"unsupported type changes ${testCase.fromType.sql} -> ${testCase.toType.sql}, " +
+      s"partitioned=$partitioned") {
+      if (partitioned) {
+        // The table needs to have at least 1 non-partition column, use a dummy one.
+        append(testCase.initialValuesDF.withColumn("dummy", lit(1)), partitionBy = Seq("value"))
+      } else {
+        append(testCase.initialValuesDF)
+      }
+      sql(s"ALTER TABLE delta.`$tempPath` " +
+        s"SET TBLPROPERTIES('delta.feature.timestampNtz' = 'supported')")
+
+      val alterTableSql =
+        s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN value TYPE ${testCase.toType.sql}"
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(alterTableSql)
+        },
+        errorClass = "NOT_SUPPORTED_CHANGE_COLUMN",
+        sqlState = None,
+        parameters = Map(
+          "table" -> s"`spark_catalog`.`delta`.`$tempPath`",
+          "originName" -> toSQLId("value"),
+          "originType" -> toSQLType(testCase.fromType),
+          "newName" -> toSQLId("value"),
+          "newType" -> toSQLType(testCase.toType)),
+        context = ExpectedContext(
+          fragment = alterTableSql,
+          start = 0,
+          stop = alterTableSql.length - 1)
+      )
+    }
+  }
+
+  test("type widening in nested fields") {
+    sql(s"CREATE TABLE delta.`$tempPath` " +
+      "(s struct<a: byte>, m map<byte, short>, a array<short>) USING DELTA")
+    append(Seq((1, 2, 3, 4))
+      .toDF("a", "b", "c", "d")
+      .selectExpr(
+        "named_struct('a', cast(a as byte)) as s",
+        "map(cast(b as byte), cast(c as short)) as m",
+        "array(cast(d as short)) as a"))
+
+    assert(readDeltaTable(tempPath).schema === new StructType()
+      .add("s", new StructType().add("a", ByteType))
+      .add("m", MapType(ByteType, ShortType))
+      .add("a", ArrayType(ShortType)))
+
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN s.a TYPE short")
```

Comment:
Can you also change the type of a STRUCT column as a whole, e.g. CHANGE COLUMN s TYPE STRUCT<...> etc.? We should make sure that it either doesn't work, or we support it correctly if that induces type widenings.

Reply:
You can't use `CHANGE COLUMN` on anything but a leaf field, I added a test to cover it (I believe it's already covered in `AlterTableTests` but it makes it to cover it here)